{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==> Total 15mer vaccine peptides: 36\n",
      "==> 29/36 with H2-b predicted ligands\n",
      "==> 22/36 with H2-d predicted ligands\n",
      "==> 29/36 with any murine predicted ligands\n",
      "==> Total 21mer vaccine peptides: 22\n",
      "==> 17/22 with H2-b predicted ligands\n",
      "==> 13/22 with H2-d predicted ligands\n",
      "==> 18/22 with any murine predicted ligands\n",
      "==> Total 27mer vaccine peptides: 19\n",
      "==> 17/19 with H2-b predicted ligands\n",
      "==> 14/19 with H2-d predicted ligands\n",
      "==> 17/19 with any murine predicted ligands\n"
     ]
    }
   ],
   "source": [
    "\n",
    "from glob import glob\n",
    "from collections import defaultdict\n",
    "\n",
    "def parse_fasta(filename):\n",
    "    result = {}\n",
    "    with open(filename) as f:\n",
    "        current_id = None\n",
    "        lines = []\n",
    "        for l in f:\n",
    "            l = l.strip()\n",
    "            if not l:\n",
    "                continue\n",
    "            elif l.startswith(\">\"):\n",
    "                if current_id:\n",
    "                    result[current_id] = \"\".join(lines)\n",
    "                current_id = l[1:].split(\"|\")[0]\n",
    "                lines = []\n",
    "            else:\n",
    "                lines.append(l)\n",
    "        if current_id:\n",
    "            result[current_id] = \"\".join(lines)\n",
    "        return result\n",
    "\n",
    "proteome = parse_fasta(\"../Figures/COVID/AA_sequence.txt\")\n",
    "\n",
    "def sort_key(f):\n",
    "    num_attributes = sum([\n",
    "        0,\n",
    "        \"h2b\" in f,\n",
    "        \"h2d\" in f,\n",
    "\n",
    "    ])\n",
    "    return (\"bcell\" in f), (\"cd4\" in f), (\"cd8\" in f), num_attributes\n",
    "\n",
    "def save_dataframe_as_fasta(df, prefix):\n",
    "    with open(prefix + \".fa\", \"w\") as f:\n",
    "        for seq, protein, start in zip(df.Sequence, df.Protein, df.Start):\n",
    "            f.write(\">SARS2-%s-%d-%d-%dmer\\n%s\\n\" % (\n",
    "                protein,\n",
    "                start,\n",
    "                start + len(seq) - 1,\n",
    "                len(seq),\n",
    "                seq,\n",
    "            ))\n",
    "            \n",
    "            \n",
    "def filename_to_symbol(source):\n",
    "    bcell = \"bcell\" in source\n",
    "    cd4 = \"cd4\" in source\n",
    "    cd8 = \"cd8\" in source\n",
    "    h2b  = \"h2b\" in source\n",
    "    h2d = \"h2d\" in source\n",
    "\n",
    "    if bcell:\n",
    "        if cd4 and cd8:\n",
    "             c = \"\\\\setlength{\\\\fboxsep}{0.5pt} \\\\boxed{\\\\circledast}\"\n",
    "               \n",
    "        elif cd4:\n",
    "            c = \"\\\\boxcircle\"\n",
    "        elif cd8:\n",
    "            c = \"\\\\boxast\"\n",
    "        else:\n",
    "            c = \"\\\\boxempty\"\n",
    "    else:\n",
    "        if cd4 and cd8:\n",
    "            c = \"\\\\circledast\"\n",
    "        elif cd4:\n",
    "            c = \"\\\\circ\"\n",
    "        elif cd8:\n",
    "            c = \"\\\\ast\"\n",
    "        else:\n",
    "            raise ValueError(\"Expected either CD4 or CD8\")\n",
    "\n",
    "    if h2b and h2d:\n",
    "        c += \"^{bd}\"\n",
    "    elif h2b:\n",
    "        c += \"^b\"\n",
    "    elif h2d:\n",
    "        c += \"^d\"\n",
    "    return c\n",
    "    \n",
    "def sources_to_hieroglyphics(sources_string):\n",
    "    sources = sources_string.split(\";\")\n",
    "    characters = []\n",
    "    for source in sorted(sources, key=sort_key):\n",
    "        characters.append(filename_to_symbol(source))\n",
    "    if len(characters) > 4:\n",
    "        symbol_string = \"\\Centerstack{ \"\n",
    "        i_values = range(len(characters) // 4 + 1)\n",
    "        for i in i_values:\n",
    "            subchars = characters[i * 4: (i + 1) * 4]\n",
    "            if subchars:\n",
    "                symbol_string += \" $%s$\" % (\" \".join(subchars))\n",
    "            if i != i_values[-1]:\n",
    "                symbol_string += \" \\\\\\\\ \"\n",
    "        symbol_string += \" }\"\n",
    "    else:\n",
    "        symbol_string = \"$ %s $\" % (\" \".join(characters))\n",
    "        \n",
    "    return symbol_string\n",
    "    \n",
    "\n",
    "def format_epitope(s):\n",
    "    if not s:\n",
    "        return \"\"\n",
    "    start = proteome[\"S\"].index(s) + 1\n",
    "    end = proteome[\"S\"].index(s) + len(s)\n",
    "    return \"S$_{%d-%d}$\" % (start, end)\n",
    "    # return \"\\\\texttt{{\\\\scriptsize %d-}%s{\\\\scriptsize -%d}}\" % (\n",
    "    #    start,\n",
    "    #    s,\n",
    "    #    end\n",
    "    #)\n",
    "\n",
    "def save_dataframe_as_latex_table(df, filename):\n",
    "    df[\"Selection Sets\"] = df.sources\n",
    "\n",
    "    # prepare table for LaTeX\n",
    "    del df[\"sources\"]\n",
    "    del df[\"HLA-I alleles\"] \n",
    "    del df[\"HLA-II alleles\"]\n",
    "    del df[\"num_sources\"]\n",
    "    class1_cov_column = \"\\Centerstack{HLA-I \\\\\\\\ Coverage}\"\n",
    "    class2_cov_column = \"\\Centerstack{HLA-II \\\\\\\\ Coverage}\"\n",
    "    df = df.rename(columns={\n",
    "        \"HLA-I coverage\": class1_cov_column,\n",
    "        \"HLA-II coverage\": class2_cov_column,\n",
    "        \"Class I H2-b\": \"H2$^{b}$ I\", \n",
    "        \"Class II H2-b\": \"H2$^{b}$ II\", \n",
    "        \"Class I H2-d\": \"H2$^{d}$ I\", \n",
    "        \"Class II H2-d\": \"H2$^{d}$ II\", \n",
    "        #\"B-cell Epitope\": \"B-cell Epitope Region\",\n",
    "        \n",
    "    })\n",
    "    df = df.reset_index()\n",
    "    \n",
    "    df.index = df.index + 1\n",
    "    with pd.option_context(\"max_colwidth\", 1000):\n",
    "        df.to_latex(\n",
    "            filename,\n",
    "            formatters={\n",
    "                \"Selection Sets\": sources_to_hieroglyphics,\n",
    "                \"Sequence\": lambda s: \"\\\\texttt{%s}\" % s,\n",
    "                \"B-cell Epitope\": format_epitope,\n",
    "                class1_cov_column: lambda x: \"%0.1f\\\\%%\" % (100 * x),\n",
    "                class2_cov_column: lambda x: \"%0.1f\\\\%%\" % (100 * x),\n",
    "            },\n",
    "            column_format=\"r\" + \"c\" * len(df.columns),\n",
    "            escape=False,\n",
    "        )\n",
    "    \n",
    "def save_dataframe(df, prefix):\n",
    "    df = df.copy()\n",
    "    df[\"has_bcell\"] = df[\"B-cell Epitope\"].str.len() > 0\n",
    "    df = df.sort_values([ \"has_bcell\", \"Protein\", \"Start\", ], ascending=True)\n",
    "    del df[\"has_bcell\"]\n",
    "    save_dataframe_as_fasta(df, prefix)\n",
    "    df = df.set_index(\"Sequence\")\n",
    "    df.to_csv(prefix + \".csv\")\n",
    "    df.to_html(prefix + \".html\")\n",
    "    save_dataframe_as_latex_table(df,  prefix + \".tex\")\n",
    "    \n",
    "\n",
    "\n",
    "    \n",
    "\n",
    "for length in [15, 21, 27]:\n",
    "    dataframes = []\n",
    "    seen_peptides = set()\n",
    "    peptide_to_sources = defaultdict(list)\n",
    "    for f in reversed(sorted(glob(\"selected-*cell*%dmer.csv\" % length), key=sort_key,)):\n",
    "        # if \"cd4\" not in f and \"cd8\" not in f:\n",
    "        #    continue\n",
    "        df_full = pd.read_csv(f)\n",
    "        peptides = df_full.Sequence.values\n",
    "        seen_already = [p in seen_peptides for p in peptides]\n",
    "        n_old = sum(seen_already)\n",
    "        n_new = len(df_full) - n_old\n",
    "        # print(\"%s: %d peptides (%d new)\" % (f, len(df_full), n_new))\n",
    "\n",
    "        for p in peptides:\n",
    "            peptide_to_sources[p].append(f)\n",
    "\n",
    "        df_sub = df_full[[\"Sequence\", \"Protein\", \"Start\", \"End\"]].copy()\n",
    "        df_sub[\"B-cell Epitope\"] = [x if type(x) is str else \"\" for x in df_full[\"full_bcell_epitope\"]]\n",
    "        df_sub[\"HLA-I alleles\"] = [x.replace(\",\", \" \").replace(\"HLA-\", \"\") if type(x) is str else \"\" for x in df_full[\"HLA-I_haplotypes\"]]\n",
    "        df_sub[\"HLA-II alleles\"] = [x.replace(\",\", \" \").replace(\"HLA-\", \"\") if type(x) is str else \"\" for x in df_full[\"HLA-II_haplotypes\"]]\n",
    "\n",
    "        df_sub[\"HLA-I coverage\"] = df_full['HLA-I_pop_freq_rounded']\n",
    "        df_sub[\"HLA-II coverage\"] = df_full['HLA-II_pop_freq_rounded']\n",
    "        df_sub['Class I H2-b'] = [\"+\" if x else \"-\" for x in df_full[\"has mouse MHC-I b\"]]\n",
    "        df_sub['Class II H2-b'] = [\"+\" if x else \"-\" for x in df_full[\"has mouse MHC-II b\"]]\n",
    "        df_sub['Class I H2-d'] = [\"+\" if x else \"-\" for x in df_full[\"has mouse MHC-I d\"]]\n",
    "        df_sub['Class II H2-d'] = [\"+\" if x else \"-\" for x in df_full[\"has mouse MHC-II d\"]]\n",
    "\n",
    "        seen_peptides.update(peptides)\n",
    "        dataframes.append(df_sub[~np.array(seen_already)])\n",
    "    df = pd.concat(dataframes); \n",
    "    df[\"sources\"] = [\";\".join(peptide_to_sources[p]) for p in df.Sequence]\n",
    "    df[\"num_sources\"] = [len(peptide_to_sources[p]) for p in df.Sequence]\n",
    "    \n",
    "    print(\"==> Total %dmer vaccine peptides: %d\" % (length, len(df),))\n",
    "    save_dataframe (df, \"final-vaccine-peptides-%dmer\" % length)\n",
    "    \n",
    "    h2b_mask = (df[\"Class I H2-b\"] == \"+\") | (df[\"Class II H2-b\"] == \"+\")\n",
    "    df_h2b = df[h2b_mask]\n",
    "    print(\"==> %d/%d with H2-b predicted ligands\" % (len(df_h2b), len(df)))\n",
    "    save_dataframe(df_h2b, \"final-vaccine-peptides-h2b-%dmer\" % length)\n",
    "    \n",
    "    h2d_mask = (df[\"Class I H2-d\"] == \"+\") | (df[\"Class II H2-d\"] == \"+\")\n",
    "    df_h2d = df[h2d_mask]\n",
    "    print(\"==> %d/%d with H2-d predicted ligands\" % (len(df_h2d), len(df)))\n",
    "    save_dataframe(df_h2d, \"final-vaccine-peptides-h2d-%dmer\" % length)\n",
    "    \n",
    "    murine_mask = h2d_mask | h2b_mask\n",
    "    df_murine = df[murine_mask]\n",
    "    print(\"==> %d/%d with any murine predicted ligands\" % (len(df_murine), len(df)))\n",
    "    save_dataframe(df_murine, \"final-vaccine-peptides-murine-%dmer\" % length)\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Protein</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>B-cell Epitope</th>\n",
       "      <th>HLA-I alleles</th>\n",
       "      <th>HLA-II alleles</th>\n",
       "      <th>HLA-I coverage</th>\n",
       "      <th>HLA-II coverage</th>\n",
       "      <th>Class I H2-b</th>\n",
       "      <th>Class II H2-b</th>\n",
       "      <th>Class I H2-d</th>\n",
       "      <th>Class II H2-d</th>\n",
       "      <th>sources</th>\n",
       "      <th>num_sources</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>YLYRLFRKSNLKPFERDISTEIYQAGS</td>\n",
       "      <td>S</td>\n",
       "      <td>451</td>\n",
       "      <td>477</td>\n",
       "      <td>FRKSNLKPFERDISTEIY</td>\n",
       "      <td>C*07:02 B*07:02 A*11:01 B*08:01 A*03:01 C*06:0...</td>\n",
       "      <td>DRB1*04:01 DRB1*13:01 DRB1*11:01 DRB1*07:01</td>\n",
       "      <td>0.78</td>\n",
       "      <td>0.46</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>selected-bcell-cd4-cd8-27mer.csv;selected-bcel...</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PQTLEILDITPCSFGGVSVITPGTNTS</td>\n",
       "      <td>S</td>\n",
       "      <td>579</td>\n",
       "      <td>605</td>\n",
       "      <td>QTLE</td>\n",
       "      <td>C*05:01</td>\n",
       "      <td>DQA1*05:05/DQB1*03:01</td>\n",
       "      <td>0.13</td>\n",
       "      <td>0.21</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>selected-bcell-cd4-cd8-27mer.csv;selected-bcel...</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PSKPSKRSFIEDLLFNKVTLADAGFIK</td>\n",
       "      <td>S</td>\n",
       "      <td>809</td>\n",
       "      <td>835</td>\n",
       "      <td>PSKP</td>\n",
       "      <td>A*02:01 A*11:01 B*08:01 A*03:01</td>\n",
       "      <td>DRB1*13:01 DRB1*01:01 DRB1*07:01</td>\n",
       "      <td>0.66</td>\n",
       "      <td>0.40</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>selected-bcell-cd4-cd8-27mer.csv;selected-bcel...</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>FRKSNLKPFERDISTEIYQAGSTPCNG</td>\n",
       "      <td>S</td>\n",
       "      <td>456</td>\n",
       "      <td>482</td>\n",
       "      <td>FRKSNLKPFERDISTEIY</td>\n",
       "      <td>C*07:02 B*07:02 A*11:01</td>\n",
       "      <td>DQA1*05:05/DQB1*03:01 DRB1*04:01</td>\n",
       "      <td>0.46</td>\n",
       "      <td>0.30</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>selected-bcell-cd4-h2b-27mer.csv</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>IYKTPPIKDFGGFNFSQILPDPSKPSK</td>\n",
       "      <td>S</td>\n",
       "      <td>788</td>\n",
       "      <td>814</td>\n",
       "      <td>PSKP</td>\n",
       "      <td>A*03:01 A*24:02</td>\n",
       "      <td>DRB1*04:01 DRB1*01:01</td>\n",
       "      <td>0.35</td>\n",
       "      <td>0.23</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>selected-bcell-cd4-h2b-27mer.csv;selected-bcel...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GNYNYLYRLFRKSNLKPFERDISTEIY</td>\n",
       "      <td>S</td>\n",
       "      <td>447</td>\n",
       "      <td>473</td>\n",
       "      <td>FRKSNLKPFERDISTEIY</td>\n",
       "      <td>C*07:02 A*24:02 B*07:02 A*11:01 B*08:01 A*03:0...</td>\n",
       "      <td>DRB1*11:01 DRB1*13:01 DRB1*07:01</td>\n",
       "      <td>0.82</td>\n",
       "      <td>0.38</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>selected-bcell-cd8-h2b-h2d-27mer.csv;selected-...</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>KFLPFQQFGRDIADTTDAVRDPQTLEI</td>\n",
       "      <td>S</td>\n",
       "      <td>558</td>\n",
       "      <td>584</td>\n",
       "      <td>QTLE</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>selected-bcell-27mer.csv</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GLTVLPPLLTDEMIAQYTSALLAGTIT</td>\n",
       "      <td>S</td>\n",
       "      <td>857</td>\n",
       "      <td>883</td>\n",
       "      <td></td>\n",
       "      <td>A*02:01 C*03:04 A*01:01 B*35:01</td>\n",
       "      <td>DQA1*05:05/DQB1*03:01 DRB1*01:01 DRB1*07:01 DR...</td>\n",
       "      <td>0.66</td>\n",
       "      <td>0.73</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>selected-tcell-cd4-cd8-h2b-h2d-27mer.csv;selec...</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FAYTKRNVIPTITQMNLKYAISAKNRA</td>\n",
       "      <td>orf1ab</td>\n",
       "      <td>4920</td>\n",
       "      <td>4946</td>\n",
       "      <td></td>\n",
       "      <td>A*01:01 B*08:01 C*03:04 C*06:02 C*07:01 C*07:0...</td>\n",
       "      <td>DRB1*01:01 DRB1*04:01 DRB1*07:01 DRB1*11:01 DR...</td>\n",
       "      <td>0.86</td>\n",
       "      <td>0.62</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>selected-tcell-cd4-cd8-h2b-h2d-27mer.csv;selec...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TLMIERFVSLAIDAYPLTKHPNQEYAD</td>\n",
       "      <td>orf1ab</td>\n",
       "      <td>5245</td>\n",
       "      <td>5271</td>\n",
       "      <td></td>\n",
       "      <td>A*03:01 A*11:01 A*02:01 B*08:01 C*03:04 C*07:0...</td>\n",
       "      <td>DQA1*05:01/DQB1*02:01 DRB1*03:01 DRB1*04:01 DR...</td>\n",
       "      <td>0.81</td>\n",
       "      <td>0.71</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>selected-tcell-cd4-cd8-h2b-h2d-27mer.csv;selec...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>IASEFSSLPSYAAFATAQEAYEQAVAN</td>\n",
       "      <td>orf1ab</td>\n",
       "      <td>3944</td>\n",
       "      <td>3970</td>\n",
       "      <td></td>\n",
       "      <td>A*01:01 B*44:02 B*44:03 C*03:04</td>\n",
       "      <td>DQA1*05:05/DQB1*03:01 DRB1*01:01 DRB1*04:01 DR...</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.84</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>selected-tcell-cd4-cd8-h2d-27mer.csv</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NYDLSVVNARLRAKHYVYIGDPAQLPA</td>\n",
       "      <td>orf1ab</td>\n",
       "      <td>5705</td>\n",
       "      <td>5731</td>\n",
       "      <td></td>\n",
       "      <td>A*03:01 A*11:01 A*24:02 C*07:02 C*06:02 C*07:0...</td>\n",
       "      <td>DRB1*13:01 DRB1*03:01 DRB1*07:01 DRB1*11:01 DR...</td>\n",
       "      <td>0.89</td>\n",
       "      <td>0.70</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>selected-tcell-cd4-cd8-h2d-27mer.csv;selected-...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DFVRATATIPIQASLPFGWLIVGVALL</td>\n",
       "      <td>ORF3a</td>\n",
       "      <td>27</td>\n",
       "      <td>53</td>\n",
       "      <td></td>\n",
       "      <td>C*03:04 B*07:02 B*35:01 A*02:01</td>\n",
       "      <td>DQA1*01:02/DQB1*06:02 DQA1*01:03/DQB1*06:03 DQ...</td>\n",
       "      <td>0.65</td>\n",
       "      <td>0.73</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>selected-tcell-cd4-cd8-h2b-27mer.csv;selected-...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>STFNVPMEKLKTLVATAEAELAKNVSL</td>\n",
       "      <td>orf1ab</td>\n",
       "      <td>2600</td>\n",
       "      <td>2626</td>\n",
       "      <td></td>\n",
       "      <td>B*44:02 B*44:03 A*11:01 C*03:04 B*08:01 A*03:0...</td>\n",
       "      <td>DQA1*01:02/DQB1*06:02 DQA1*05:01/DQB1*02:01 DQ...</td>\n",
       "      <td>0.69</td>\n",
       "      <td>0.73</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>selected-tcell-cd4-cd8-27mer.csv</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ATPSDFVRATATIPIQASLPFGWLIVG</td>\n",
       "      <td>ORF3a</td>\n",
       "      <td>23</td>\n",
       "      <td>49</td>\n",
       "      <td></td>\n",
       "      <td>C*03:04 B*07:02 B*35:01</td>\n",
       "      <td>DRB1*01:01 DRB1*04:01 DRB1*07:01 DQA1*01:02/DQ...</td>\n",
       "      <td>0.39</td>\n",
       "      <td>0.78</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>selected-tcell-cd4-h2b-h2d-27mer.csv;selected-...</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RLQSLQTYVTQQLIRAAEIRASANLAA</td>\n",
       "      <td>S</td>\n",
       "      <td>1000</td>\n",
       "      <td>1026</td>\n",
       "      <td></td>\n",
       "      <td>A*02:01 B*44:02 B*44:03</td>\n",
       "      <td>DRB1*01:01 DRB1*13:01 DRB1*15:01 DRB1*04:01 DR...</td>\n",
       "      <td>0.54</td>\n",
       "      <td>0.81</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>selected-tcell-cd4-h2b-h2d-27mer.csv;selected-...</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRLSFKELLVYAADPAMHAASGNLLLD</td>\n",
       "      <td>orf1ab</td>\n",
       "      <td>4756</td>\n",
       "      <td>4782</td>\n",
       "      <td></td>\n",
       "      <td>A*01:01 B*35:01 A*03:01 C*03:04 A*02:01 C*06:0...</td>\n",
       "      <td>DRB1*01:01 DQA1*05:05/DQB1*03:01 DRB1*03:01 DR...</td>\n",
       "      <td>0.87</td>\n",
       "      <td>0.80</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>selected-tcell-cd4-h2b-h2d-27mer.csv;selected-...</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>LLQFAYANRNRFLYIIKLIFLWLLWPV</td>\n",
       "      <td>M</td>\n",
       "      <td>34</td>\n",
       "      <td>60</td>\n",
       "      <td></td>\n",
       "      <td>B*08:01 C*07:02 B*35:01 C*03:04 C*06:02 C*07:0...</td>\n",
       "      <td>DRB1*13:01 DRB1*03:01 DRB1*11:01</td>\n",
       "      <td>0.89</td>\n",
       "      <td>0.36</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>selected-tcell-cd8-h2b-h2d-27mer.csv;selected-...</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HLDGEVITFDNLKTLLSLREVRTIKVF</td>\n",
       "      <td>orf1ab</td>\n",
       "      <td>1545</td>\n",
       "      <td>1571</td>\n",
       "      <td></td>\n",
       "      <td>A*11:01 C*05:01 C*03:04 B*08:01 B*44:02 B*44:0...</td>\n",
       "      <td>DRB1*04:01 DRB1*15:01 DRB1*01:01 DRB1*11:01 DR...</td>\n",
       "      <td>0.74</td>\n",
       "      <td>0.62</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>selected-tcell-cd8-h2b-h2d-27mer.csv;selected-...</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      Sequence Protein  Start   End      B-cell Epitope  \\\n",
       "0  YLYRLFRKSNLKPFERDISTEIYQAGS       S    451   477  FRKSNLKPFERDISTEIY   \n",
       "1  PQTLEILDITPCSFGGVSVITPGTNTS       S    579   605                QTLE   \n",
       "2  PSKPSKRSFIEDLLFNKVTLADAGFIK       S    809   835                PSKP   \n",
       "0  FRKSNLKPFERDISTEIYQAGSTPCNG       S    456   482  FRKSNLKPFERDISTEIY   \n",
       "1  IYKTPPIKDFGGFNFSQILPDPSKPSK       S    788   814                PSKP   \n",
       "0  GNYNYLYRLFRKSNLKPFERDISTEIY       S    447   473  FRKSNLKPFERDISTEIY   \n",
       "1  KFLPFQQFGRDIADTTDAVRDPQTLEI       S    558   584                QTLE   \n",
       "0  GLTVLPPLLTDEMIAQYTSALLAGTIT       S    857   883                       \n",
       "1  FAYTKRNVIPTITQMNLKYAISAKNRA  orf1ab   4920  4946                       \n",
       "2  TLMIERFVSLAIDAYPLTKHPNQEYAD  orf1ab   5245  5271                       \n",
       "1  IASEFSSLPSYAAFATAQEAYEQAVAN  orf1ab   3944  3970                       \n",
       "2  NYDLSVVNARLRAKHYVYIGDPAQLPA  orf1ab   5705  5731                       \n",
       "0  DFVRATATIPIQASLPFGWLIVGVALL   ORF3a     27    53                       \n",
       "2  STFNVPMEKLKTLVATAEAELAKNVSL  orf1ab   2600  2626                       \n",
       "0  ATPSDFVRATATIPIQASLPFGWLIVG   ORF3a     23    49                       \n",
       "1  RLQSLQTYVTQQLIRAAEIRASANLAA       S   1000  1026                       \n",
       "2  SRLSFKELLVYAADPAMHAASGNLLLD  orf1ab   4756  4782                       \n",
       "0  LLQFAYANRNRFLYIIKLIFLWLLWPV       M     34    60                       \n",
       "1  HLDGEVITFDNLKTLLSLREVRTIKVF  orf1ab   1545  1571                       \n",
       "\n",
       "                                       HLA-I alleles  \\\n",
       "0  C*07:02 B*07:02 A*11:01 B*08:01 A*03:01 C*06:0...   \n",
       "1                                            C*05:01   \n",
       "2                    A*02:01 A*11:01 B*08:01 A*03:01   \n",
       "0                            C*07:02 B*07:02 A*11:01   \n",
       "1                                    A*03:01 A*24:02   \n",
       "0  C*07:02 A*24:02 B*07:02 A*11:01 B*08:01 A*03:0...   \n",
       "1                                                      \n",
       "0                    A*02:01 C*03:04 A*01:01 B*35:01   \n",
       "1  A*01:01 B*08:01 C*03:04 C*06:02 C*07:01 C*07:0...   \n",
       "2  A*03:01 A*11:01 A*02:01 B*08:01 C*03:04 C*07:0...   \n",
       "1                    A*01:01 B*44:02 B*44:03 C*03:04   \n",
       "2  A*03:01 A*11:01 A*24:02 C*07:02 C*06:02 C*07:0...   \n",
       "0                    C*03:04 B*07:02 B*35:01 A*02:01   \n",
       "2  B*44:02 B*44:03 A*11:01 C*03:04 B*08:01 A*03:0...   \n",
       "0                            C*03:04 B*07:02 B*35:01   \n",
       "1                            A*02:01 B*44:02 B*44:03   \n",
       "2  A*01:01 B*35:01 A*03:01 C*03:04 A*02:01 C*06:0...   \n",
       "0  B*08:01 C*07:02 B*35:01 C*03:04 C*06:02 C*07:0...   \n",
       "1  A*11:01 C*05:01 C*03:04 B*08:01 B*44:02 B*44:0...   \n",
       "\n",
       "                                      HLA-II alleles  HLA-I coverage  \\\n",
       "0        DRB1*04:01 DRB1*13:01 DRB1*11:01 DRB1*07:01            0.78   \n",
       "1                              DQA1*05:05/DQB1*03:01            0.13   \n",
       "2                   DRB1*13:01 DRB1*01:01 DRB1*07:01            0.66   \n",
       "0                   DQA1*05:05/DQB1*03:01 DRB1*04:01            0.46   \n",
       "1                              DRB1*04:01 DRB1*01:01            0.35   \n",
       "0                   DRB1*11:01 DRB1*13:01 DRB1*07:01            0.82   \n",
       "1                                                               0.00   \n",
       "0  DQA1*05:05/DQB1*03:01 DRB1*01:01 DRB1*07:01 DR...            0.66   \n",
       "1  DRB1*01:01 DRB1*04:01 DRB1*07:01 DRB1*11:01 DR...            0.86   \n",
       "2  DQA1*05:01/DQB1*02:01 DRB1*03:01 DRB1*04:01 DR...            0.81   \n",
       "1  DQA1*05:05/DQB1*03:01 DRB1*01:01 DRB1*04:01 DR...            0.48   \n",
       "2  DRB1*13:01 DRB1*03:01 DRB1*07:01 DRB1*11:01 DR...            0.89   \n",
       "0  DQA1*01:02/DQB1*06:02 DQA1*01:03/DQB1*06:03 DQ...            0.65   \n",
       "2  DQA1*01:02/DQB1*06:02 DQA1*05:01/DQB1*02:01 DQ...            0.69   \n",
       "0  DRB1*01:01 DRB1*04:01 DRB1*07:01 DQA1*01:02/DQ...            0.39   \n",
       "1  DRB1*01:01 DRB1*13:01 DRB1*15:01 DRB1*04:01 DR...            0.54   \n",
       "2  DRB1*01:01 DQA1*05:05/DQB1*03:01 DRB1*03:01 DR...            0.87   \n",
       "0                   DRB1*13:01 DRB1*03:01 DRB1*11:01            0.89   \n",
       "1  DRB1*04:01 DRB1*15:01 DRB1*01:01 DRB1*11:01 DR...            0.74   \n",
       "\n",
       "   HLA-II coverage Class I H2-b Class II H2-b Class I H2-d Class II H2-d  \\\n",
       "0             0.46            +             -            -             -   \n",
       "1             0.21            -             -            -             -   \n",
       "2             0.40            +             -            -             +   \n",
       "0             0.30            -             +            -             -   \n",
       "1             0.23            -             +            -             -   \n",
       "0             0.38            +             -            +             -   \n",
       "1             0.00            -             -            -             -   \n",
       "0             0.73            +             +            +             +   \n",
       "1             0.62            +             +            +             +   \n",
       "2             0.71            +             +            +             +   \n",
       "1             0.84            +             +            +             +   \n",
       "2             0.70            -             +            +             +   \n",
       "0             0.73            +             +            -             +   \n",
       "2             0.73            -             +            -             +   \n",
       "0             0.78            +             +            -             +   \n",
       "1             0.81            -             +            +             +   \n",
       "2             0.80            -             +            +             +   \n",
       "0             0.36            +             +            +             +   \n",
       "1             0.62            +             -            +             +   \n",
       "\n",
       "                                             sources  num_sources  \n",
       "0  selected-bcell-cd4-cd8-27mer.csv;selected-bcel...            3  \n",
       "1  selected-bcell-cd4-cd8-27mer.csv;selected-bcel...            3  \n",
       "2  selected-bcell-cd4-cd8-27mer.csv;selected-bcel...            5  \n",
       "0                   selected-bcell-cd4-h2b-27mer.csv            1  \n",
       "1  selected-bcell-cd4-h2b-27mer.csv;selected-bcel...            2  \n",
       "0  selected-bcell-cd8-h2b-h2d-27mer.csv;selected-...            4  \n",
       "1                           selected-bcell-27mer.csv            1  \n",
       "0  selected-tcell-cd4-cd8-h2b-h2d-27mer.csv;selec...            4  \n",
       "1  selected-tcell-cd4-cd8-h2b-h2d-27mer.csv;selec...            2  \n",
       "2  selected-tcell-cd4-cd8-h2b-h2d-27mer.csv;selec...            2  \n",
       "1               selected-tcell-cd4-cd8-h2d-27mer.csv            1  \n",
       "2  selected-tcell-cd4-cd8-h2d-27mer.csv;selected-...            2  \n",
       "0  selected-tcell-cd4-cd8-h2b-27mer.csv;selected-...            2  \n",
       "2                   selected-tcell-cd4-cd8-27mer.csv            1  \n",
       "0  selected-tcell-cd4-h2b-h2d-27mer.csv;selected-...            4  \n",
       "1  selected-tcell-cd4-h2b-h2d-27mer.csv;selected-...            4  \n",
       "2  selected-tcell-cd4-h2b-h2d-27mer.csv;selected-...            4  \n",
       "0  selected-tcell-cd8-h2b-h2d-27mer.csv;selected-...            4  \n",
       "1  selected-tcell-cd8-h2b-h2d-27mer.csv;selected-...            4  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "old_peptides = set([l.strip() for l in open(\"order-old-sorted.txt\")])\n",
    "new_peptides = set([l.strip() for l in open(\"order-new-sorted.txt\")])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'GAAAYYVGYLQPRTFLLKYNENGTITD', 'SETKCTLKSFTVEKGIYQTSNFRVQPT'}"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "{p for p in new_peptides if p not in old_peptides}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'GKGYHLMSFPQSAPHGVVFLHVTYVPA',\n",
       " 'LQPELDSFKEELDKYFKNHTSPDVDLG',\n",
       " 'SLIDLQELGKYEQYIKWPWYIWLGFIA',\n",
       " 'TFVSGNCDVVIGIVNNTVYDPLQPELD',\n",
       " 'VYDPLQPELDSFKEELDKYFKNHTSPD'}"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "{p for p in old_peptides if p not in new_peptides}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This is XeTeX, Version 3.14159265-2.6-0.999992 (TeX Live 2020) (preloaded format=xelatex)\r\n",
      " restricted \\write18 enabled.\r\n",
      "entering extended mode\r\n",
      "(./tables.tex\r\n",
      "LaTeX2e <2020-02-02> patch level 5\r\n",
      "L3 programming layer <2020-03-06>\r\n",
      "(/usr/local/texlive/2020/texmf-dist/tex/latex/base/article.cls\r\n",
      "Document Class: article 2019/12/20 v1.4l Standard LaTeX document class\r\n",
      "(/usr/local/texlive/2020/texmf-dist/tex/latex/base/size10.clo))\r\n",
      "(/usr/local/texlive/2020/texmf-dist/tex/latex/geometry/geometry.sty\r\n",
      "(/usr/local/texlive/2020/texmf-dist/tex/latex/graphics/keyval.sty)\r\n",
      "(/usr/local/texlive/2020/texmf-dist/tex/generic/iftex/ifvtex.sty\r\n",
      "(/usr/local/texlive/2020/texmf-dist/tex/generic/iftex/iftex.sty)))\r\n",
      "(/usr/local/texlive/2020/texmf-dist/tex/latex/base/inputenc.sty\r\n",
      "\r\n",
      "Package inputenc Warning: inputenc package ignored with utf8 based engines.\r\n",
      "\r\n",
      ") (/usr/local/texlive/2020/texmf-dist/tex/latex/graphics/color.sty\r\n",
      "(/usr/local/texlive/2020/texmf-dist/tex/latex/graphics-cfg/color.cfg)\r\n",
      "(/usr/local/texlive/2020/texmf-dist/tex/latex/graphics-def/xetex.def))\r\n",
      "(/usr/local/texlive/2020/texmf-dist/tex/latex/colortbl/colortbl.sty\r\n",
      "(/usr/local/texlive/2020/texmf-dist/tex/latex/tools/array.sty))\r\n",
      "(/usr/local/texlive/2020/texmf-dist/tex/latex/amsfonts/amssymb.sty\r\n",
      "(/usr/local/texlive/2020/texmf-dist/tex/latex/amsfonts/amsfonts.sty))\r\n",
      "(/usr/local/texlive/2020/texmf-dist/tex/latex/amsmath/amsmath.sty\r\n",
      "For additional information on amsmath, use the `?' option.\r\n",
      "(/usr/local/texlive/2020/texmf-dist/tex/latex/amsmath/amstext.sty\r\n",
      "(/usr/local/texlive/2020/texmf-dist/tex/latex/amsmath/amsgen.sty))\r\n",
      "(/usr/local/texlive/2020/texmf-dist/tex/latex/amsmath/amsbsy.sty)\r\n",
      "(/usr/local/texlive/2020/texmf-dist/tex/latex/amsmath/amsopn.sty))\r\n",
      "(/usr/local/texlive/2020/texmf-dist/tex/latex/stmaryrd/stmaryrd.sty)\r\n",
      "(/usr/local/texlive/2020/texmf-dist/tex/latex/stackengine/stackengine.sty\r\n",
      "(/usr/local/texlive/2020/texmf-dist/tex/latex/etoolbox/etoolbox.sty)\r\n",
      "(/usr/local/texlive/2020/texmf-dist/tex/generic/listofitems/listofitems.sty\r\n",
      "(/usr/local/texlive/2020/texmf-dist/tex/generic/listofitems/listofitems.tex))\r\n",
      "(/usr/local/texlive/2020/texmf-dist/tex/latex/tools/calc.sty))\r\n",
      "(/usr/local/texlive/2020/texmf-dist/tex/latex/booktabs/booktabs.sty)\r\n",
      "(/usr/local/texlive/2020/texmf-dist/tex/latex/l3backend/l3backend-xdvipdfmx.def\r\n",
      ") (./tables.aux) (/usr/local/texlive/2020/texmf-dist/tex/latex/base/ts1cmr.fd)\r\n",
      "*geometry* driver: auto-detecting\r\n",
      "*geometry* detected driver: xetex\r\n",
      "(./summary-27mer.tex\r\n",
      "(/usr/local/texlive/2020/texmf-dist/tex/latex/amsfonts/umsa.fd)\r\n",
      "(/usr/local/texlive/2020/texmf-dist/tex/latex/amsfonts/umsb.fd)\r\n",
      "(/usr/local/texlive/2020/texmf-dist/tex/latex/stmaryrd/Ustmry.fd))\r\n",
      "(./final-vaccine-peptides-27mer.tex) [1] (./summary-21mer.tex) [2]\r\n",
      "(./final-vaccine-peptides-21mer.tex) [3] (./summary-15mer.tex) [4]\r\n",
      "(./final-vaccine-peptides-15mer.tex) [5]\r\n",
      "Overfull \\vbox (7.00243pt too high) has occurred while \\output is active\r\n",
      "[6] (./tables.aux) )\r\n",
      "(see the transcript file for additional information)\r\n",
      "Output written on tables.pdf (6 pages).\r\n",
      "Transcript written on tables.log.\r\n"
     ]
    }
   ],
   "source": [
    "!xelatex tables.tex"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "!open tables.pdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
